src/textconv.c

```diff
@@ -0,0 +1,611 @@
+static int textconv_registry_insert(
+                                  const char *name, git_textconv *textconv)
+{
+    git_textconv_entry *entry;
+    
+    size_t alloc_len = 0;
+    GITERR_CHECK_ALLOC_ADD(&alloc_len, alloc_len, sizeof(git_textconv_entry));
+    
+    entry = git__calloc(1, alloc_len);
+    GITERR_CHECK_ALLOC(entry);
+    
+    entry->textconv_name = git__strdup(name);
+    GITERR_CHECK_ALLOC(entry->textconv_name);
+    
+    entry->textconv      = textconv;
+    
+    if (git_vector_insert(&textconv_registry.textconvs, entry) < 0) {
+        git__free(entry->textconv_name);
+        git__free(entry);
+        return -1;
+    }
+    
+    return 0;
+}
+
+int git_textconv_global_init(void)
+{
+    git_textconv *http = NULL;
+    int error = 0;
+    
+    if (git_rwlock_init(&textconv_registry.lock) < 0)
+        return -1;
+    
+    if ((error = git_vector_init(&textconv_registry.textconvs, 2,
+                                 textconv_entry_name_cmp)) < 0)
+        goto done;
+    
+    git__on_shutdown(git_textconv_global_shutdown);
+    
+done:
+    if (error) {
+        git_textconv_free(http);
+    }
+    
+    return error;
+}
+
+static void git_textconv_global_shutdown(void)
+{
+    size_t pos;
+    git_textconv_entry *entry;
+    
+    if (git_rwlock_wrlock(&textconv_registry.lock) < 0)
+        return;
+    
+    git_vector_foreach(&textconv_registry.textconvs, pos, entry) {
+        if (entry->textconv && entry->textconv->shutdown) {
+            entry->textconv->shutdown(entry->textconv);
+            entry->initialized = false;
+        }
+        
+        git__free(entry->textconv_name);
+        git__free(entry);
+    }
+    
+    git_vector_free(&textconv_registry.textconvs);
+    
+    git_rwlock_wrunlock(&textconv_registry.lock);
+    git_rwlock_free(&textconv_registry.lock);
+}
+
+/* Note: callers must lock the registry before calling this function */
+static int textconv_registry_find(size_t *pos, const char *name)
+{
+    return git_vector_bsearch2(pos, &textconv_registry.textconvs, textconv_entry_name_key_check, name);
+}
+
+/* Note: callers must lock the registry before calling this function */
+static git_textconv_entry *textconv_registry_lookup(size_t *pos, const char *name)
+{
+    git_textconv_entry *entry = NULL;
+    
+    if (!textconv_registry_find(pos, name))
+        entry = git_vector_get(&textconv_registry.textconvs, *pos);
+    
+    return entry;
+}
+
+
+int git_textconv_register(
+                        const char *name, git_textconv *textconv)
+{
+    int error;
+    
+    assert(name && textconv);
+    
+    if (git_rwlock_wrlock(&textconv_registry.lock) < 0) {
+        giterr_set(GITERR_OS, "failed to lock textconv registry");
+        return -1;
+    }
+    
+    if (!textconv_registry_find(NULL, name)) {
+        giterr_set(
+                   GITERR_TEXTCONV, "attempt to reregister existing textconv '%s'", name);
```

Comment:
Style: strange line split. Our limit is "soft", so if you're only off by a few characters from 80, don't bother.